fix: fancy indexing fixes backed h5py error

sjfleming opened the pull request:
<!-- Please:
1. Fill in the following check boxes
2. Make sure checks pass (Ignore “Triage” ones)
-->

- [x] Closes #2064 
- [x] Tests added
- [ ] Release note added (or unnecessary)

This is a first stab at fixing #2064 by adding `_safe_fancy_index_h5py` (and 3 related helper functions) to `anndata/_core/index.py`.  The function `_safe_fancy_index_h5py` only gets called in the case where there are repeated indices being requested (this is the only case that is currently causing a bug, so in all other cases, the existing code -- `d[tuple(ordered)][tuple(rev_order)]` -- is what runs).

sjfleming commented on src/anndata/_core/index.py:
```diff
@@ -224,10 +224,91 @@ def _subset_dataset(d, subset_idx):
             ordered[axis] = axis_idx[order]
             rev_order[axis] = np.argsort(order)
+    # check for duplicates or multi-dimensional fancy indexing
+    has_duplicates = False
+    num_array_dims = 0
+    for axis_idx in ordered:
+        if isinstance(axis_idx, np.ndarray):
+            num_array_dims += 1
+            if len(np.unique(axis_idx)) != len(axis_idx):
+                has_duplicates = True
+                break
```
Thanks

sjfleming commented on src/anndata/_core/index.py:
```diff
@@ -217,10 +217,91 @@ def _subset_dataset(d: h5py.Dataset, subset_idx: Index):
+        return result[tuple(slices)]
+
+
+def _safe_fancy_index_h5py(dataset, subset_idx):
+    # Handle multi-dimensional indexing while being memory-efficient
+    # This avoids h5py's limitation with multi-dimensional fancy indexing
+    # without loading the entire dataset into memory
+    if not isinstance(subset_idx, tuple):
+        subset_idx = (subset_idx,)
```
Yes

sjfleming commented on src/anndata/_core/index.py:
```diff
@@ -217,10 +217,91 @@ def _subset_dataset(d: h5py.Dataset, subset_idx: Index):
+        subset_idx = (subset_idx,)
+
+    # Convert boolean arrays to integer arrays and handle sorting for h5py
+    processed_indices = []
+    reverse_indices = []
+
+    for idx in subset_idx:
+        processed_idx, reverse_idx = _process_index_for_h5py(idx)
+        processed_indices.append(processed_idx)
+        reverse_indices.append(reverse_idx)
```
Looks good

sjfleming commented:
I took a shot at it.  Let me know what you think @flying-sheep !  Thanks

sjfleming commented:
Thanks @flying-sheep !  Yes I think your type changes make sense.  Please check and see whether my comments and changes address your concerns (which identified a problem with my previous type-hints for `_apply_index_to_result`)

sjfleming commented:
Hi @ilan-gold , I will look into making a benchmark.  You mean creating a new benchmark to cover these changes, right?  I didn't see an existing benchmark that looked like it would cover this use.

So I think that, the way this PR is written, if the requested indices have no repeats, then `_subset_dataset` performs (one or two) additional call(s) to `np.unique` here
https://github.com/sjfleming/anndata/blob/bdab9001bd096a63b4a6129e02b36af23d47046d/src/anndata/_core/index.py#L231

but if there are no repeats (which is probably almost always the case) the code skips the new `_safe_fancy_index_h5py` function altogether and just returns what it used to return:
https://github.com/sjfleming/anndata/blob/bdab9001bd096a63b4a6129e02b36af23d47046d/src/anndata/_core/index.py#L238

sjfleming commented:
Currently this seems like a major problem, so I'm trying to track down what's going on

```
| Change   | Before [c9c0be49] <main>   | After [2d5efbc3] <sf-backed-hdf5-fancy-indexing>   | Ratio   | Benchmark (Parameter)                                                                               |
|----------|----------------------------|----------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------|
|          | 8.015625                   | 8.296875                                           | 1.04    | readwrite.H5ADWriteSuite.track_peakmem_write_compressed('pbmc3k')                                   |
| +        | 0.96875                    | 9.359375                                           | 9.66    | readwrite.H5ADWriteSuite.track_peakmem_write_full('pbmc3k')                                         |
```

ilan-gold commented:
@sjfleming So the benchmarks literally crashed here, not sure why!

Since the becnhmark took 174 minutes to run, could you either (a) enusre things run locally first but not in 174 minutes or (b) let me know that you want to use the CI for sanity checks and then ping me when you want the "real" benchmarks run on our machine? Then I'll remove the label and put it back when ready.  Thanks a bunch!

sjfleming commented:
@ilan-gold okay so the benchmarks I messed with in my attempt to "improve stability" [9734678](https://github.com/scverse/anndata/pull/2066/commits/973467864a06cb19c24220146bf5058e66d4332f) are now failing to return a value, so that was ill-conceived.  (It worked locally...)
Those are the ones with prefix `readwrite.....track_peakmem...` that now have all zeros.  I will revert this change.  It seems like I was misunderstanding how that memory calc worked: when I run on my laptop, I guess I'm seeing memory effects from a bunch of other processes.  When it runs on a dedicated machine in CI, it's probably fine.

There are new benchmarks I've written in `benchmarks/benchmarks/backed_hdf5.py` (`BackedHDF5.....timed_fancy_index...`) that will fail on `main` without this PR, because they cover new functionality.
Maybe it's a bad idea to include things like that?  Not sure what you think.

I can run the full benchmarking suite locally with `asv run --python same`. The benchmarks I've added take about 22 mins to run. The full run of all benchmarks (including those new ones) completes in 25 mins and everything looks okay.  Things also look good compared to `main` if I run on main and then do `asv compare`.

But it looks like I've added too much to the benchmarking.  I'm going to cut down on the number of new ones so as not to bog down the benchmarking suite.  I think I went way overboard.

sjfleming commented:
I've pared down the new benchmarks so that they run in 2 additional minutes.

ilan-gold commented:
@sjfleming I actually just noticed something: https://github.com/scverse/anndata/pull/2066/files#diff-4e5148021dbb054cace3f9749e52227b27fa61acf1137d979f90d148b82555c3R15-R21 only contains sparse benchmarking but this indexing is for dense, no?

If that is true, I would revert this PR (https://github.com/scverse/anndata/pull/2163) and then we would benchmark against main again with the fixed benchmarks.  But maybe I'm missing something.

Alternatively, we could do this manually locally, but I'd kind of rather have the "clean" run on the benchmark machine.  Let me know what you think
